qiime/core/path.py

```diff
@@ -0,0 +1,46 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016--, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import pathlib
+import shutil
+import tempfile
+import typing
+
+
+T = typing.TypeVar('T')
+_ConcretePath = type(pathlib.Path())
+
+
+class TempPath(_ConcretePath, typing.Generic[T]):
+    def __new__(cls, dir=False, **kwargs):
+        """
+        Create a tempfile, return pathlib.Path reference to it.
+        """
+        if dir:
+            name = tempfile.mkdtemp(**kwargs)
+        else:
+            _, name = tempfile.mkstemp(**kwargs)
+        return super().__new__(cls, name)
+
+    def __del__(self):
+        self._finalize()
+        if hasattr(super(), "__del__"):
+            super().__del__(self)
+
+    def __exit__(self, t, v, tb):
+        self._finalize()
+        super().__exit__(t, v, tb)
+
+    def _finalize(self):
+        if not self.exists():
+            return
+
+        if self.is_dir():
+            shutil.rmtree(str(self))
+        else:
+            self.unlink()
```

Comment:
I see, the context manager adds an idea of opening and closing "paths" (not
files), so unlink isn't called anyways in the parent-classes' exit.

On Mon, Aug 22, 2016 at 1:16 PM, Matthew Dillon <notifications@github.com>
wrote:

> In qiime/core/path.py
> <https://github.com/qiime2/qiime2/pull/117#discussion_r75749923>:
>
> > +        self._finalize()
> > +        if hasattr(super(), "__del__"):
> > +            super().__del__(self)
> > +
> > +    def __exit__(self, t, v, tb):
> > +        self._finalize()
> > +        super().__exit__(t, v, tb)
> > +
> > +    def _finalize(self):
> > +        if not self.exists():
> > +            return
> > +
> > +        if self.is_dir():
> > +            shutil.rmtree(str(self))
> > +        else:
> > +            self.unlink()
>
> No, Path.unlink is not idempotent. Path.unlink removes files/links at
> time of execution, and Path.rmdir removes empty directories. I decided to
> use shutil.rmtree to handle removing non-empty dirs.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/qiime2/qiime2/pull/117/files/d40a7679783d676af6bdb09da9b6ff351d2f87dc#r75749923>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADyuZM2YNqARde-si7R2t36JidEoQHqgks5qigOjgaJpZM4JqQCg>
> .
>